tests/test_docs_extraction.py

```diff
@@ -0,0 +1,334 @@
+
+    @dec_noop
+    class MyModel(BaseModel):
+        a: int
+        """A docs"""
+
+        model_config = ConfigDict(
+            use_attribute_docstrings=True,
+        )
+
+    @dec_noop
+    class MyModel(BaseModel):
+        b: int
+        """B docs"""
+
+        model_config = ConfigDict(
+            use_attribute_docstrings=True,
+        )
+
+    assert MyModel.model_fields['b'].description == 'B docs'
+
+    # With https://github.com/python/cpython/pull/106815/ introduced,
+    # inspect will fallback to the last found class in the source file.
+    # The following is to ensure using frames will still get the correct one
+    if True:
+
+        class MyModel(BaseModel):
+            a: int
+            """A docs"""
+
+            model_config = ConfigDict(
+                use_attribute_docstrings=True,
+            )
+
+    else:
+
+        class MyModel(BaseModel):
+            b: int
+            """B docs"""
+
+            model_config = ConfigDict(
+                use_attribute_docstrings=True,
+            )
+
+    assert MyModel.model_fields['a'].description == 'A docs'
+
+
+def test_model_docs_dedented_string():
+    # fmt: off
+    class MyModel(BaseModel):
+        def bar(self):
+            """
+An inconveniently dedented string
+            """
+
+        a: int
+        """A docs"""
+
+        model_config = ConfigDict(
+            use_attribute_docstrings=True,
+        )
+    # fmt: on
+    assert MyModel.model_fields['a'].description == 'A docs'
+
+
+def test_model_docs_inheritance():
+    class MyModel(BaseModel):
+        a: int
+        """A docs"""
+
+        b: int
+        """B docs"""
+
+        model_config = ConfigDict(
+            use_attribute_docstrings=True,
+        )
+
+    FirstModel = MyModel
+
+    class MyModel(FirstModel):
+        a: int
+        """A overridden docs"""
+
+    assert FirstModel.model_fields['a'].description == 'A docs'
+    assert MyModel.model_fields['a'].description == 'A overridden docs'
+    assert MyModel.model_fields['b'].description == 'B docs'
+
+
+def test_model_different_name():
+    # As we extract docstrings from cls in `ModelMetaclass.__new__`,
+    # we are not affected by `__name__` being altered in any way.
+
+    class MyModel(BaseModel):
+        a: int
+        """A docs"""
+
+        model_config = ConfigDict(
+            use_attribute_docstrings=True,
+        )
+
+    MyModel.__name__ = 'OtherModel'
+    print(MyModel.__name__)
+
+    assert MyModel.model_fields['a'].description == 'A docs'
+
+
+def test_model_generic():
+    class MyModel(BaseModel, Generic[T]):
+        a: T
+        """A docs"""
+
+        model_config = ConfigDict(
+            use_attribute_docstrings=True,
+        )
+
+    assert MyModel.model_fields['a'].description == 'A docs'
+
+    class MyParameterizedModel(MyModel[int]):
+        a: int
+        """A parameterized docs"""
+
+    assert MyParameterizedModel.model_fields['a'].description == 'A parameterized docs'
```

Comment:
We hit an interesting behavior here. If you try to debug with a breakpoint L51 in `_docs_extraction.py`, you will see `_extract_source_from_frame` is called two times.

`BaseModel.__class_getitem__` will manually return a new class object, and will use `model_parametrized_name` to set the name of this new class. So when writing `class MyParameterizedModel(MyModel[int])`, a first class is created when `MyModel[int]` is executed (we can't find any class declaration because the model name is set to `MyModel[int]` by Pydantic,  so no docs are extracted). Then when `MyParameterizedModel(...)` is executed and docs are successfully extracted here.

Reply:
I was worried that the user overriding `model_parametrized_name` might be a problem, but it seems to be fine:

```python
import typing
from typing import Generic

from pydantic import BaseModel, ConfigDict

T = typing.TypeVar("T")


class MyModel(BaseModel, Generic[T]):
    a: T
    """A docs"""

    model_config = ConfigDict(
        use_attribute_docstrings=True,
    )

    @classmethod
    def model_parametrized_name(cls, params):
        return "OtherName"


assert MyModel.model_fields["a"].description == "A docs"
assert MyModel[int].model_fields["a"].description == "A docs"  # maybe add this to the test
assert MyModel[int].__name__ == "OtherName"
assert MyModel.__name__ == "MyModel"


class MyParameterizedModel(MyModel[int]):
    a: int
    """A parameterized docs"""


assert MyParameterizedModel.model_fields["a"].description == "A parameterized docs"
assert MyParameterizedModel.__name__ == "MyParameterizedModel"
```